src/Tools/dotnet-dsrouter/DiagnosticsServerRouterCommands.cs

```diff
@@ -45,254 +45,274 @@ public void OnRouterStopped()
-                }
+                Launcher.SuspendProcess = true;
+                Launcher.ConnectMode = false;
+                Launcher.Verbose = LogLevel != LogLevel.Information;
+                Launcher.CommandToken = cancellationToken;
             }
+        }
 
-            if (string.IsNullOrEmpty(ipcServer))
-                ipcServer = GetDefaultIpcServerPath(logger);
+        public async Task<int> RunIpcClientTcpClientRouter(CancellationToken token, string ipcClient, string tcpClient, int runtimeTimeout, string verbose, string forwardPort)
+        {
+            var runner = new IpcClientTcpClientRunner(verbose);
+            return await runner.CommonRunLoop((logger, launcherCallbacks, linkedCancelToken) =>
+            {
+                TcpClientRouterFactory.CreateInstanceDelegate tcpClientRouterFactory = ChooseTcpClientRouterFactory(forwardPort, logger);
 
-            var routerTask = DiagnosticsServerRouterRunner.runIpcServerTcpClientRouter(linkedCancelToken.Token, ipcServer, tcpClient, runtimeTimeout == Timeout.Infinite ? runtimeTimeout : runtimeTimeout * 1000, tcpClientRouterFactory, logger, Launcher);
+                var routerTask = DiagnosticsServerRouterRunner.runIpcClientTcpClientRouter(linkedCancelToken.Token, ipcClient, tcpClient, runtimeTimeout == Timeout.Infinite ? runtimeTimeout : runtimeTimeout * 1000, tcpClientRouterFactory, logger, launcherCallbacks);
+                return routerTask;
+            }, token);
+        }
 
-            while (!linkedCancelToken.IsCancellationRequested)
-            {
-                await Task.WhenAny(routerTask, Task.Delay(250)).ConfigureAwait(false);
-                if (routerTask.IsCompleted)
-                    break;
+        class IpcServerWebSocketServerRunner : SpecificRunnerBase
+        {
+            public IpcServerWebSocketServerRunner(string verbose) : base(verbose) { }
 
-                if (!Console.IsInputRedirected && Console.KeyAvailable)
-                {
-                    ConsoleKey cmd = Console.ReadKey(true).Key;
-                    if (cmd == ConsoleKey.Q)
-                    {
-                        cancelRouterTask.Cancel();
-                        break;
-                    }
-                }
+            public override void ConfigureLauncher(CancellationToken cancellationToken)
+            {
+                Launcher.SuspendProcess = false;
+                Launcher.ConnectMode = true;
+                Launcher.Verbose = LogLevel != LogLevel.Information;
+                Launcher.CommandToken = cancellationToken;
             }
-
-            return routerTask.Result;
         }
 
-        public async Task<int> RunIpcClientTcpClientRouter(CancellationToken token, string ipcClient, string tcpClient, int runtimeTimeout, string verbose, string forwardPort)
+        public async Task<int> RunIpcServerWebSocketServerRouter(CancellationToken token, string ipcServer, string webSocket, int runtimeTimeout, string verbose)
```

Comment:
I guess this should use  checkLoopbackOnly as well, we need to warn if you try to bind anything except loopback interface.